docs/src/03_data.md

```diff
@@ -0,0 +1,97 @@
+# Dataflow and the `Data` trait
+
+The Druid architecture is based on a two-way dataflow.
+
+At the root level, you define the application state, which is passed to each child widget as associated data. Some Widgets (eg LensWrap) will only pass a subset of that data to their children.
+
+Some widgets (eg Button, TextBox, Checkbox) can mutate the data passed to them by their parents in reaction to user events. The data mutated in a child widget is also changed in the parent widgets, all the way to the root.
+
+When you mutate a widget's associated data, Druid compares the old and new version, and propagates the change to the widgets that are affected by the change.
+
+Note that, in all that workflow, Widgets don't actually store their associated data. A `Button<Vector<String>>` doesn't actually store a `Vector<String>`, the framework just stores one which is provided to the button.
+
+For this to work, your model must implement the `Clone` and `Data` traits. The `Data` trait has a single method:
+
+```rust,no_run,noplaypen
+{{#include ../../druid/src/data.rs:same_fn}}
+```
+
+This method checks for equality, but allows for false negatives.
+
+
+## Performance
+
+It is important that your data be cheap to clone and cheap to compare; we encourage the use of reference counted pointers to allow cheap cloning of more expensive types. `Arc` and `Rc` have blanket `Data` impls that do pointer comparison, so if you have a type that does not implement `Data`, you can always just wrap it in one of those smart pointers.
```

Comment:
```suggestion
It is important that your data is cheap to clone and cheap to compare; we encourage the use of reference counted pointers to allow cheap cloning of more expensive types. `Arc` and `Rc` have blanket `Data` impls that do pointer comparison, so if you have a type that does not implement `Data`, you can always just wrap it in one of those smart pointers.
```

Alternatively *are* if data is plural, but I think that would confuse more people so I went with singular data.